### PR TITLE
[Compute] (1) vm create: vm create: Add --vmss to specify an existing virtual machine scale set that the virtual machine should be assigned to. (2) vmss create: Add --orchestrator to specify how virtual machines are managed by the scale set.

### PR DESCRIPTION
**Pull request by qwordy**
* vm create: Add --vmss to specify an existing virtual machine scale set that the virtual machine should be assigned to.
* vmss create: Add --orchestrator to specify how virtual machines are managed by the scale set.

Implement #10156 

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by qwordy** [comment]
Testing has some problems now.

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py [hunk 1]
```diff
@@ -373,6 +373,9 @@ def _build_storage_profile():
     if availability_set_id:
         vm_properties['availabilitySet'] = {'id': availability_set_id}
 
+    if vmss is not None:
+        vm_properties['virtualMachineScaleSet'] = {'id': vmss}
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/vm/_template_builder.py]: Looks like you just support `ID`

File: src/azure-cli/azure/cli/command_modules/vm/_params.py [hunk 2]
```diff
@@ -276,6 +276,7 @@ def load_arguments(self, _):
         c.argument('vm_name', name_arg_type, id_part=None, help='Name of the virtual machine.', completer=None)
         c.argument('os_disk_size_gb', type=int, help='the size of the os disk in GB', arg_group='Storage')
         c.argument('availability_set', help='Name or ID of an existing availability set to add the VM to. None by default.')
+        c.argument('vmss', help='Virtual machine scale set that the virtual machine should be assigned to.')
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/vm/_params.py]: Name or ID of Virtual machine scale set that the virtual machine should be assigned to.

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py (hunk 1 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/_template_builder.py]: > 
> 
> Looks like you just support `ID`

I copy the above line. I have ask this in email.

File: src/azure-cli/azure/cli/command_modules/vm/_params.py (hunk 2 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/_params.py]: > 
> 
> Name or ID of Virtual machine scale set that the virtual machine should be assigned to.

Good

**Comment by yonzhan** [comment]
@qwordy  what is the status of this PR right now ?

**Comment by qwordy** [comment]
> 
> 
> @qwordy what is the status of this PR right now ?

Use case problem. 
And after many PR were merged, conflicts sprang up.
I need to fix both of them.

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py (hunk 1 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/_template_builder.py]: I have supported name now.

**Comment by qwordy** [comment]
@shpa-microsoft, please help review. Thanks.

**Comment by qwordy** [comment]
> 
> 
> > @qwordy what is the status of this PR right now ?
> 
> Use case problem.
> And after many PR were merged, conflicts sprang up.
> I need to fix both of them.

Solved.

File: src/azure-cli/HISTORY.rst [hunk 3]
```diff
@@ -25,6 +25,8 @@ Release History
 **Compute**
 
 * vm create: Add warning when specifying accelerated networking and an existing NIC together.
+* vm create: Add --vmss to specify virtual machine scale set that the virtual machine should be assigned to.
+* vmss create: Add --empty to create a VMSS without VM profile, which can be then referenced in `vm create`.
```
Comment (shpa-microsoft) [inline src/azure-cli/HISTORY.rst]: Empty keyword will give a wrong indication IMO cause one can create an empty VMSS even when VM profile exists. We have started to use orchestrationMode to specify this. The 2 modes are: VMSS and VM. I'd also seek feedback from Jun and Varun.

Comment (shpa-microsoft) [reply src/azure-cli/HISTORY.rst]: Also vmss create should happen before vm create in this case. Not sure ordering matters in the doc.

File: src/azure-cli/azure/cli/command_modules/vm/_params.py [hunk 4]
```diff
@@ -276,6 +276,7 @@ def load_arguments(self, _):
         c.argument('vm_name', name_arg_type, id_part=None, help='Name of the virtual machine.', completer=None)
         c.argument('os_disk_size_gb', type=int, help='the size of the os disk in GB', arg_group='Storage')
         c.argument('availability_set', help='Name or ID of an existing availability set to add the VM to. None by default.')
+        c.argument('vmss', help='Name of ID of a virtual machine scale set that the virtual machine should be assigned to. None by default.')
```
Comment (shpa-microsoft) [inline src/azure-cli/azure/cli/command_modules/vm/_params.py]: Just like that mentioned in avset. We need to specify "existing"

File: src/azure-cli/azure/cli/command_modules/vm/_params.py [hunk 5]
```diff
@@ -496,6 +497,7 @@ def load_arguments(self, _):
                    help="The eviction policy for virtual machines in a low priority scale set.", is_preview=True)
         c.argument('application_security_groups', resource_type=ResourceType.MGMT_COMPUTE, min_api='2018-06-01', nargs='+', options_list=['--asgs'], help='Space-separated list of existing application security groups to associate with the VM.', arg_group='Network', validator=validate_asg_names_or_ids)
         c.argument('computer_name_prefix', help='Computer name prefix for all of the virtual machines in the scale set. Computer name prefixes must be 1 to 15 characters long')
+        c.argument('empty', help='Create a VMSS without VM profile, which can be then referenced in `vm create`.', arg_type=get_three_state_flag())
```
Comment (shpa-microsoft) [inline src/azure-cli/azure/cli/command_modules/vm/_params.py]: Nit: Creates 
Additional: If we change the term to orchestration mode, then all this description will change.

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py [hunk 6]
```diff
@@ -387,6 +387,10 @@ def _build_storage_profile():
     if availability_set_id:
         vm_properties['availabilitySet'] = {'id': availability_set_id}
 
+    # vmss is ID
+    if vmss is not None:
+        vm_properties['virtualMachineScaleSet'] = {'id': vmss}
```
Comment (shpa-microsoft) [inline src/azure-cli/azure/cli/command_modules/vm/_template_builder.py]: How are we supporting just the name? We need full ARM ID of VMSS when it hits the rest layer. So if you guys are converting that to ARM ID, this is fine. Or else follow what avset does.

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py [hunk 7]
```diff
@@ -847,6 +851,22 @@ def build_vmss_resource(cmd, name, naming_prefix, location, tags, overprovision,
     }
     if zones:
         vmss['zones'] = zones
+    # vmss without vm profile
+    if empty:
+        if platform_fault_domain_count is None:
+            platform_fault_domain_count = 2
```
Comment (shpa-microsoft) [inline src/azure-cli/azure/cli/command_modules/vm/_template_builder.py]: Why is it defaulted to 2? Can this not be passed down as null?

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py (hunk 6 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/_template_builder.py]: I handle this in _validate_vm_create_vmss(). This function can transform name to ID. I wrote examples in test cases.

File: src/azure-cli/HISTORY.rst (hunk 3 above)
Comment (qwordy) [reply src/azure-cli/HISTORY.rst]: > 
> 
> Also vmss create should happen before vm create in this case. Not sure ordering matters in the doc.

This doesn't matter. It's just change log.

Comment (qwordy) [reply src/azure-cli/HISTORY.rst]: > 
> 
> Empty keyword will give a wrong indication IMO cause one can create an empty VMSS even when VM profile exists. We have started to use orchestrationMode to specify this. The 2 modes are: VMSS and VM. I'd also seek feedback from Jun and Varun.

I think "empty" is ambiguous, too. Do you have a suggested name?
Is there a change of vm and vmss creation ways in the future? You mentioned orchestrationMode.

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py (hunk 7 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/_template_builder.py]: When I use null,
Deployment failed. Correlation ID: e7f42bc9-2d4b-4310-90cd-8291da117c5d. {
  "error": {
    "code": "InvalidParameter",
    "message": "Required parameter 'platformFaultDomainCount' is missing (null).",
    "target": "platformFaultDomainCount"
  }
}

File: src/azure-cli/azure/cli/command_modules/vm/_params.py (hunk 5 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/_params.py]: When will you change?

File: src/azure-cli/azure/cli/command_modules/vm/_params.py (hunk 4 above)
Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/vm/_params.py]: Added

**Comment by qwordy** [comment]
Another question, after I create a vmss without vm profile, I can't run commands like list-instances, get-instance-view on it,
```
Operation 'VirtualMachineScaleSets.virtualMachines.GET' is not allowed on Virtual Machine Scale Set 'vmss1'
```
```
Operation 'VirtualMachineScaleSets.instanceView.GET' is not allowed on Virtual Machine Scale Set 'vmss1'.
```
Is it normal?

File: src/azure-cli/azure/cli/command_modules/vm/_params.py (hunk 5 above)
Comment (shpa-microsoft) [reply src/azure-cli/azure/cli/command_modules/vm/_params.py]: We have that term on portal already. Let's have a call today and clarify all the gaps?

File: src/azure-cli/azure/cli/command_modules/vm/_template_builder.py (hunk 7 above)
Comment (shpa-microsoft) [reply src/azure-cli/azure/cli/command_modules/vm/_template_builder.py]: Oh My bad. We do need this one; but not sure defaulting is the right thing to do v/s making clients always enter that value.

**Comment by qwordy** [comment]
Shalaka and I have discussed in a phone call.

File: src/azure-cli/azure/cli/command_modules/vm/_params.py (hunk 5 above)
Comment (shpa-microsoft) [reply src/azure-cli/azure/cli/command_modules/vm/_params.py]: Decided on enum "orchestration" for this purpose. Keeping it in sync with Portal

File: src/azure-cli/azure/cli/command_modules/vm/_params.py [hunk 8]
```diff
@@ -276,6 +276,7 @@ def load_arguments(self, _):
         c.argument('vm_name', name_arg_type, id_part=None, help='Name of the virtual machine.', completer=None)
         c.argument('os_disk_size_gb', type=int, help='the size of the os disk in GB', arg_group='Storage')
         c.argument('availability_set', help='Name or ID of an existing availability set to add the VM to. None by default.')
+        c.argument('vmss', help='Name of ID of an existing virtual machine scale set that the virtual machine should be assigned to. None by default.')
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/vm/_params.py]: Name or ID of

**Comment by shpa-microsoft** [comment]
Can we name the param: **OrchestrationMode** instead? With options "VM" and "ScaleSetVM"

**Comment by qwordy** [comment]
> 
> 
> Can we name the param: **OrchestrationMode** instead? With options "VM" and "ScaleSetVM"

I've merged it to enable related people to try it. I will open another PR for this problem.

**Pull request merged**